### PR TITLE
Fix for broken_promise exception observed on connection shutdown

### PR DESCRIPTION
**Pull request by graphcareful**
REVIEWERS: Please review https://github.com/redpanda-data/seastar/pull/115 instead

A broken_promise exception was observed when calling `close()` due to the fact that the input stream was called while it was waiting on a pending operation to complete.

This occurred on call to close because of how `do_shutdown()` is implemented. `do_shutdown()` acquires the output semaphore, but not the input, then attempts to `wait_for_input()`. If there is already a waiter on the read path, then the broken_promise exception will be observed when `do_shutdown()` calls `do_get()`.

The fix is to simply modify `do_shutdown()` to call `wait_for_eof()` instead of wait_for_input(),however this would be incomplete because there are other instances in the implementation where out of band operations on the read / write path come from opposite paths.

To resolve the above concern new methods have been developed (maybe_wait_for_data and maybe_send_data) which attempt to acquire the in/out semaphore before performing the desired operation. If the semaphore is already acquired, then the method just returns to the caller, which would lead to a retry until the operation can be performed.

Fixes: CORE-539

### CODE REVIEW

**Comment by graphcareful** [comment]
Changes in [force-push](https://github.com/redpanda-data/seastar/compare/b140183938d8caf7bb29470a3a0978a5f121608c..c3a5aca9b8f09ec1a187e45f5c576a3b6fc29939)
- Split formatting commit out

**Comment by graphcareful** [comment]
Changes in [force-push](https://github.com/redpanda-data/seastar/compare/403239d24f7fb04dcee86fe1f731ad5f35be6ef2..5c76161326ce7e8bc982c9bb1fbd6b5ceea9c86b)
- Removed accidental duplicate call wrapping `do_get()` in context of semaphore
- Slightly reworked logic in `wait_for_eof()` , previously this depended on waiting on `_in_sem` a deadlock was detected in a unit test. Its been since refactored to use `wait_for_input()` instead

File: src/net/ossl.cc [hunk 1]
```diff
@@ -717,6 +722,23 @@ class session : public enable_shared_from_this<session>, public session_impl {
...
 
+    // Call this method when its desired to have a read performed, but if a read is already being
+    // performed, just return when that read completes
+    future<> wait_for_data() {
+        auto avail = BIO_pending(_in_bio);
+        auto units = try_get_units(_in_sem, 1);
+        if (units) {
+            units->return_all();
+            return wait_for_input();
+        }
```
Comment (michael-redpanda) [inline src/net/ossl.cc:732]: I'm a little confused by this.  If you get units, you immediately return then and then wait on the input (`_in`) stream.  Why don't you reutrn the units after the return from wait_for_input?

File: src/net/ossl.cc [hunk 2]
```diff
@@ -564,7 +564,7 @@ class session : public enable_shared_from_this<session>, public session_impl {
             std::unique_ptr<net::connected_socket_impl> sock, sstring name = { }, std::optional<tls_options> options = std::nullopt)
             : _type(t), _sock(std::move(sock)), _creds(creds->_impl), _hostname(
                    std::move(name)), _in(_sock->source()), _out(_sock->sink()),
-                   _in_sem(1), _out_sem(1),  _options(options.value_or(tls_options{})),
+                   _in_sem(1), _out_sem(1), _send_sem(1), _options(options.value_or(tls_options{})),
```
Comment (michael-redpanda) [inline src/net/ossl.cc:566]: qq: why not just use `_out_sem`?  If the only time you need to write to `_out` is post `BIO_read(_out_bio)` woudln't `_out_sem` suffice?

Comment (graphcareful) [reply src/net/ossl.cc:566]: `_out_sem` is taken by `put` and `put` eventually calls `pull_encrypted_and_send`. Therefore there would be a deadlock if I were to guard `_out` with `_out_sem`

File: src/net/ossl.cc (hunk 1 above)
Comment (graphcareful) [reply src/net/ossl.cc:732]: Nice catch, this wasn't correctly updated after the latest force push

Comment (graphcareful) [reply src/net/ossl.cc:732]: On another note, not really sure how to test this edge case. It occurs when `SSL_write` gets `SSL_ERROR_WANT_READ`

**Comment by graphcareful** [comment]
Changes in [force-push](https://github.com/redpanda-data/seastar/compare/d76a31ac179688c5506cee81245b375227cdfea8..c09cd25ace24a501cf303ff42a07863ae378c130)
- Modified implementation to not use 2 additional semaphores and instead retry if semaphore is already taken , otherwise take lease and perform work.
- Ensure read lock is taken within `wait_for_eof`

File: src/net/ossl.cc [hunk 3]
```diff
@@ -1178,15 +1202,17 @@ class session : public enable_shared_from_this<session>, public session_impl {
...
+        if (!X509_NAME_print_ex(out.get(), name, 0, ASN1_STRFLGS_RFC2253 | XN_FLAG_SEP_COMMA_PLUS |
+                                XN_FLAG_FN_SN | XN_FLAG_DUMP_UNKNOWN_FIELDS)) {
+            return std::nullopt;
         }
-        return std::nullopt;
+        static const auto max_len = 4096;
+        char buffer[max_len];
+        memset(buffer, 0, max_len);
+        BIO_read(out.get(), buffer, max_len - 1);
+        return sstring(buffer);
```
Comment (michael-redpanda) [inline src/net/ossl.cc:1215]:
```suggestion
        char * bio_ptr;
        auto len = BIO_get_mem_data(out.get(), &bio_ptr);
        return sstring(bio_ptr, len);
```

**Review by michael-redpanda — COMMENTED** [review]
Just one clean up suggestion

File: src/net/ossl.cc [hunk 4]
```diff
@@ -1202,15 +1202,17 @@ class session : public enable_shared_from_this<session>, public session_impl {
...
-    static std::optional<sstring> get_ossl_string(X509_NAME* name){
-        if (auto name_str = X509_NAME_oneline(name, nullptr, 0)) {
-            // sstring constructor may throw, to ensure deallocation of this OpenSSL string in
-            // all cases, wrap the call to free() in a deferred_action
-            auto done = defer([&name_str]() noexcept { OPENSSL_free(name_str); });
-            sstring ossl_str(name_str);
-            return ossl_str;
+    static std::optional<sstring> get_dn_string(X509_NAME* name) {
+        auto out = bio_ptr(BIO_new(BIO_s_mem()));
+        if (!X509_NAME_print_ex(out.get(), name, 0, ASN1_STRFLGS_RFC2253 | XN_FLAG_SEP_COMMA_PLUS |
```
Comment (BenPope) [inline src/net/ossl.cc:1207]: > X509_NAME_print_ex() and X509_NAME_print_ex_fp() return 1 on success or 0 on error if the XN_FLAG_COMPAT is set, which is the same as X509_NAME_print(). Otherwise, it returns -1 on error or other values on success.

Should the check be `-1 == X509_NAME_print_ex`?

File: src/net/ossl.cc (hunk 3 above)
Comment (BenPope) [reply src/net/ossl.cc:1215]: >        char * bio_ptr = nullptr;
>        auto len = BIO_get_mem_data(out.get(), &bio_ptr);
>        return sstring(bio_ptr, len);

> BIO_get_mem_data() returns the total number of bytes available on success, 0 if b is NULL, or a negative value in case of other errors.

Looks like the error condition should be checked?

File: tests/unit/tls_test.cc [hunk 5]
```diff
@@ -1444,6 +1444,10 @@ SEASTAR_THREAD_TEST_CASE(test_dn_name_handling) {
         fout.get();
 
         auto dn = fdn.get();
+        BOOST_REQUIRE(dn.has_value());
```
Comment (BenPope) [inline tests/unit/tls_test.cc:1447]: Are there tests for the failure modes?

Comment (graphcareful) [reply tests/unit/tls_test.cc:1447]: Nope, i believe these unit tests could be enhanced a bit more

**Comment by graphcareful** [comment]
Note to reviewers, base has been modified to `rebase-feature-ossl` which is the `feature-ossl` branch with the seastar 24.2.x changes rebased onto it.

File: src/net/ossl.cc [hunk 6]
```diff
@@ -848,6 +848,9 @@ class session : public enable_shared_from_this<session>, public session_impl {
                 f = pull_encrypted_and_send();
             }
             return f.then([this]{
+                if (!_options.wait_for_eof_on_shutdown) {
+                    return make_ready_future<>();
+                }
```
Comment (BenPope) [inline src/net/ossl.cc]: Can you add a note about why this is "sidestepped".

File: src/net/ossl.cc [hunk 7]
```diff
@@ -315,11 +315,9 @@ class tls::certificate_credentials::impl {
         if (!pkey) {
             throw ossl_error("Error attempting to parse private key");
         }
-#if 0 // https://github.com/redpanda-data/core-internal/issues/1233
-        if (!X509_verify(x509_cert.get(), pkey.get())) {
+        if (!X509_check_private_key(x509_cert.get(), pkey.get())) {
```
Comment (BenPope) [inline src/net/ossl.cc:318]: Can a test be added?

File: src/net/ossl.cc [hunk 8]
```diff
@@ -828,7 +828,7 @@ class session : public enable_shared_from_this<session>, public session_impl {
     }
 
     future<> do_shutdown() {
-        if(_error || !connected() || eof()) {
+        if(_error || !connected()) {
```
Comment (BenPope) [inline src/net/ossl.cc:831]: Can a test be added?

File: src/net/ossl.cc [hunk 9]
```diff
@@ -1170,30 +1189,34 @@ class session : public enable_shared_from_this<session>, public session_impl {
...
+        return sstring(bio_ptr, len);
     }
 
     future<> client_handshake() {
         return do_handshake(SSL_connect, [this]{
             return pull_encrypted_and_send().then([this]{
                 return wait_for_input().then([this]{
                     if (eof()) {
-                        return make_exception_future<>(std::runtime_error("EOF observed during handshake"));
+                        _error = std::make_exception_ptr(std::system_error(ENOTCONN, std::system_category()));
```
[the remark this answers is not in the document]
Comment (BenPope) [reply src/net/ossl.cc:1218]: Is it possible to add a test?

The commit message talks about Redpanda, but I think it would be more appropriate to discuss this in terms of compatibility with existing behaviour in GnuTLS.

[the remark this answers is not in the document]
Comment (michael-redpanda) [reply src/net/ossl.cc:1218]: > The commit message talks about Redpanda, but I think it would be more appropriate to discuss this in terms of compatibility with existing behaviour in GnuTLS.

In this situation, GnuTLS will return an error code which turns into a `system_error` which Redpanda does parse to see if it's a reconnect type error.

File: cmake/SeastarDependencies.cmake [hunk 10]
```diff
@@ -130,6 +131,8 @@ macro (seastar_find_dependencies)
     VERSION 1.7.3)
   seastar_set_dep_args (GnuTLS REQUIRED
     VERSION 3.3.26)
+  seastar_set_dep_args (OpenSSL REQUIRED
+    VERSION 3.0.0)
```
[the remark this answers is not in the document]
Comment (michael-redpanda) [reply cmake/SeastarDependencies.cmake:135]: It's a minimum.

File: src/net/ossl.cc (hunk 8 above)
Comment (michael-redpanda) [reply src/net/ossl.cc:831]: There are a number of shutdown unit tests in Seastar that verify the shutdown behavior

File: src/net/ossl.cc (hunk 7 above)
Comment (michael-redpanda) [reply src/net/ossl.cc:318]: Seastar's TLS unit tests do verify the coherency of the key and cert

File: src/net/ossl.cc (hunk 6 above)
Comment (michael-redpanda) [reply src/net/ossl.cc]: This matches original behavior in the GnuTLS implementation

**Comment by michael-redpanda** [comment]
Leaving this in draft, but will close.  Please review https://github.com/redpanda-data/seastar/pull/115
